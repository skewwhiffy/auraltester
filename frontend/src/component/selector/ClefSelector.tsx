import H1 from "../H1";
import Centre from "../Centre";
import FancyRadioButtons from "../FancyRadioButtons";

export default ClefSelector

interface Props {
  title?: string | null
  value?: string | null

  onChange(value: string): void
}

function ClefSelector({title, value, onChange}: Props) {
  return (
    <>
      {title && <Centre><H1>{title}</H1></Centre>}
      <Centre>
        <FancyRadioButtons value={value} onChange={onChange}>
          <FancyRadioButtons.Item>Bass</FancyRadioButtons.Item>
        </FancyRadioButtons>
      </Centre>
    </>
  )
}
