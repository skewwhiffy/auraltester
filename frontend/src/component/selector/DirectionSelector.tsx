import Centre from "../Centre"
import FancyRadioButtons from "../FancyRadioButtons"
import H1 from "../H1"
import {Direction} from "../../common/types";

interface DirectionSelectorProps {
    title?: string | null,
    value?: Direction | null,

    onChange(direction: Direction): void
}

const DirectionSelector = ({title, value, onChange}: DirectionSelectorProps) => {
    return (
        <>
            {title && <Centre><H1>{title}</H1></Centre>}
            <Centre>
                <FancyRadioButtons value={value} onChange={onChange}>
                    <FancyRadioButtons.Item value="ascending">Ascending</FancyRadioButtons.Item>
                    <FancyRadioButtons.Item value="descending">Descending</FancyRadioButtons.Item>
                </FancyRadioButtons>
            </Centre>
        </>
    )
}

export default DirectionSelector
