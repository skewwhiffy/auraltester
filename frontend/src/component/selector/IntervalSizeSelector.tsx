import Centre from "../Centre";
import FancyRadioButtons from "../FancyRadioButtons";
import {IntervalSize} from "../../common/types";
import H1 from "../H1";

interface Props {
    title?: string
    value?: IntervalSize

    onChange(value: IntervalSize): void
}

const IntervalSizeSelector = ({title, value, onChange}: Props) => {
    return (
        <>
            {title && <Centre><H1>{title}</H1></Centre>}
            <Centre>
                <FancyRadioButtons value={value} onChange={onChange}>
                    <FancyRadioButtons.Item value="1">Unison</FancyRadioButtons.Item>
                    <FancyRadioButtons.Item value="2">Second</FancyRadioButtons.Item>
                    <FancyRadioButtons.Item value="3">Third</FancyRadioButtons.Item>
                    <FancyRadioButtons.Item value="4">Fourth</FancyRadioButtons.Item>
                    <FancyRadioButtons.Item value="5">Fifth</FancyRadioButtons.Item>
                    <FancyRadioButtons.Item value="6">Sixth</FancyRadioButtons.Item>
                    <FancyRadioButtons.Item value="7">Seventh</FancyRadioButtons.Item>
                    <FancyRadioButtons.Item value="8">Octave</FancyRadioButtons.Item>
                </FancyRadioButtons>
            </Centre>
        </>
    )
}

export default IntervalSizeSelector