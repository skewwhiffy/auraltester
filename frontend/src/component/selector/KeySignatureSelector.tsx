import React from "react"
import { Col, Container, Row } from 'react-bootstrap'
import RadioButtons, { RadioButtonDefinition } from "../../util/RadioButtons"

type OnChangeHandler = (note: string) => void

interface Props {
  defaultValue?: string
  onChange: OnChangeHandler
}

interface State {
  keySignature: string
}

class KeySignatureSelector extends React.Component<Props, State>{
  constructor(props: Props) {
    super(props)
    this.state = {
      keySignature: props.defaultValue || 'C'
    }
    props.onChange(this.state.keySignature)
  }

  private keySignatureValues: Array<RadioButtonDefinition> = [
    { value: 'C#', label: 'C# / A# minor' },
    { value: 'F#', label: 'F# / D# minor' },
    { value: 'B', label: 'B / G# minor' },
    { value: 'E', label: 'E / C# minor' },
    { value: 'A', label: 'A / F# minor' },
    { value: 'D', label: 'D / B minor' },
    { value: 'G', label: 'G / E minor' },
    { value: 'C', label: 'None / C / A minor' },
    { value: 'F', label: 'F / D minor' },
    { value: 'Bb', label: 'Bb / G minor' },
    { value: 'Eb', label: 'Eb / C minor' },
    { value: 'Ab', label: 'Ab / F minor' },
    { value: 'Db', label: 'Db / Bb minor' },
    { value: 'Gb', label: 'Gb / Eb minor' },
    { value: 'Cb', label: 'Cb / Ab minor' },
  ]

  render() {
    return (
      <Container>
        <Row><Col>Key signature</Col></Row>
        <Row>
          <Col>
            <RadioButtons
              values={this.keySignatureValues}
              name='keySignature'
              defaultValue={this.state.keySignature}
              onChange={keySignature => this.props.onChange(keySignature)}
            />
          </Col>
        </Row>
      </Container>
    )
  }
}

export default KeySignatureSelector