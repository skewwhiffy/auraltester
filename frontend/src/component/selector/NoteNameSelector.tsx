import Centre from "../Centre";
import H1 from "../H1";
import FancyRadioButtons from "../FancyRadioButtons";
import {NoteName} from "../../common/types";

interface Props {
    title?: string,
    value?: NoteName,

    onChange(value: NoteName): void
}

const NoteNameSelector = ({title, value, onChange}: Props) => {
    return (
        <>
            {title && <Centre><H1>{title}</H1></Centre>}
            <Centre>
                <FancyRadioButtons
                    value={value}
                    onChange={onChange}
                >
                    <FancyRadioButtons.Item>A</FancyRadioButtons.Item>
                    <FancyRadioButtons.Item>B</FancyRadioButtons.Item>
                    <FancyRadioButtons.Item>C</FancyRadioButtons.Item>
                    <FancyRadioButtons.Item>D</FancyRadioButtons.Item>
                    <FancyRadioButtons.Item>E</FancyRadioButtons.Item>
                    <FancyRadioButtons.Item>F</FancyRadioButtons.Item>
                    <FancyRadioButtons.Item>G</FancyRadioButtons.Item>
                </FancyRadioButtons>
            </Centre>
        </>
    )
}

export default NoteNameSelector