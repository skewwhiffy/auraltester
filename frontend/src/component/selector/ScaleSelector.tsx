import React from 'react'
import { Form, Container, Row, Col } from 'react-bootstrap'
import RadioButtons from '../../util/RadioButtons'
import ClefSelector from './ClefSelector'
import { capitalizeFirstCharacter } from '../../util'

type OnChangeHandler = (
  clef: string,
  note: string,
  type: string,
  direction: string
) => void

interface Props {
  onChange: OnChangeHandler
}

interface State {
  clef: string,
  note: string,
  accidental: string,
  type: string,
  direction: string
}

class ScaleSelector extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props)
    this.state = {
      clef: 'treble',
      note: 'C',
      accidental: '',
      type: 'major',
      direction: 'ascending'
    }
    this.onStateChange(this.state)
  }

  render() {
    return (
      <Container>
        <Form>
          <Row>
            <Col>
              <Container>
                <Row><Col>Note</Col></Row>
                <Row>
                  <Col>
                    <RadioButtons
                      values={'ABCDEFG'.split('').map(it => ({ value: it }))}
                      name='note'
                      defaultValue={this.state.note}
                      onChange={note => this.onFormChange({ ...this.state, note })}
                    />
                  </Col>
                  <Col>
                    <RadioButtons
                      values={[
                        { value: '#' },
                        { value: '', label: 'natural' },
                        { value: 'b' }
                      ]}
                      name='accidental'
                      defaultValue={this.state.accidental}
                      onChange={accidental => this.onFormChange({ ...this.state, accidental })}
                    />
                  </Col>
                </Row>
              </Container>
            </Col>
            <Col>
              <ClefSelector
                defaultValue={this.state.clef}
                onChange={clef => this.onFormChange({ ...this.state, clef })}
              />
            </Col>
            <Col>
              <RadioButtons
                values={['major', 'minor-harmonic', 'minor-melodic']
                  .map(it => ({
                    value: it,
                    label: it.split('-').map(capitalizeFirstCharacter).join(' ')
                  }))}
                name='type'
                defaultValue={this.state.type}
                onChange={type => this.onFormChange({ ...this.state, type })}
              />
            </Col>
            <Col>
              <RadioButtons
                values={['ascending', 'descending']
                  .map(it => ({
                    value: it,
                    label: capitalizeFirstCharacter(it)
                  }))}
                name='direction'
                defaultValue={this.state.direction}
                onChange={direction => this.onFormChange({ ...this.state, direction })}
              />
            </Col>
          </Row>
        </Form>
      </Container>
    )
  }

  private onFormChange = (newState: State) => {
    this.setState(newState)
    this.onStateChange(newState)
  }

  private onStateChange(state: State) {
    this.props.onChange(
      state.clef,
      `${state.note}${state.accidental}`,
      state.type,
      state.direction
    )
  }
}

export default ScaleSelector