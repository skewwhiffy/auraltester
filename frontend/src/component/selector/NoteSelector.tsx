import {useState} from "react"
import {Accidental, Note, NoteName} from "../../common/types"
import Centre from "../Centre"
import FancyRadioButtons from "../FancyRadioButtons"
import H1 from "../H1"
import NoteNameSelector from "./NoteNameSelector";

interface NoteSelectorProps {
    title?: string | null,
    value?: Note | null,
    showDoubles?: boolean

    onChange(note: Note): void
}

const NoteSelector = ({title, value, onChange, showDoubles = true}: NoteSelectorProps) => {
    const [noteName, setNoteName] = useState<NoteName | undefined>(value?.name)
    const [accidental, setAccidental] = useState<Accidental | undefined>(value?.accidental)

    const onNoteNameChange = (noteName: NoteName) => {
        setNoteName(noteName)
        if (accidental !== undefined && accidental !== null) {
            onChange({name: noteName, accidental})
        }
    }

    const onAccidentalChange = (accidental: Accidental) => {
        setAccidental(accidental)
        if (noteName !== undefined && noteName !== null) {
            onChange({name: noteName, accidental})
        }
    }

    return (
        <>
            {title && <Centre><H1>{title}</H1></Centre>}
            <NoteNameSelector value={noteName} onChange={onNoteNameChange}/>
            <Centre>
                <FancyRadioButtons
                    value={accidental}
                    onChange={onAccidentalChange}
                >
                    {showDoubles && <FancyRadioButtons.Item value='x'>Double sharp</FancyRadioButtons.Item>}
                    <FancyRadioButtons.Item>#</FancyRadioButtons.Item>
                    <FancyRadioButtons.Item value=''>Natural</FancyRadioButtons.Item>
                    <FancyRadioButtons.Item>b</FancyRadioButtons.Item>
                    {showDoubles && <FancyRadioButtons.Item value='bb'>Double flat</FancyRadioButtons.Item>}
                </FancyRadioButtons>
            </Centre>
        </>
    )
}

export default NoteSelector