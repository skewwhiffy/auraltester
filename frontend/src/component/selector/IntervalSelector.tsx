import { useState } from 'react'
import { Col, Container, Form, Row } from 'react-bootstrap'
import RadioButtons from '../../util/RadioButtons'
import ClefSelector from './ClefSelector'
import KeySignatureSelector from './KeySignatureSelector'
import NoteSelector from './NoteSelector'

type OnChangeHandler = (
  clef: string,
  bottomNote: string,
  intervalQuality: string,
  intervalSize: string,
  keySignature: string
) => void

export interface Props {
  onChange: OnChangeHandler
}

interface State {
  clef?: string
  bottomNote?: string
  intervalQuality?: string
  intervalSize?: string
  keySignature?: string
}

const IntervalSelector = (props: Props): JSX.Element => {
  const [state, setState] = useState<State>({})

  const onFormChange = (newState: State): void => {
    setState(newState)
    onStateChange(newState)
  }

  const onStateChange = (state: State): void => {
    if (state.clef === undefined ||
      state.bottomNote === undefined ||
      state.intervalQuality === undefined ||
      state.intervalSize === undefined ||
      state.keySignature === undefined) {
      return
    }
    props.onChange(
      state.clef,
      state.bottomNote,
      state.intervalQuality,
      state.intervalSize,
      state.keySignature
    )
  }

  return (
    <Container>
      <Form>
        <Row>
          <Col>
            <ClefSelector
              value={state.clef}
              onChange={clef => { onFormChange({ ...state, clef }) }}
            />
          </Col>
          <Col>
            <NoteSelector
              value={state.bottomNote}
              includeDoubleAccidentals={true}
              onChange={bottomNote => { onFormChange({ ...state, bottomNote }) }}
            />
          </Col>
          <Col>
            <Container>
              <Row><Col>Interval</Col></Row>
              <Row>
                <Col>
                  <RadioButtons
                    values={['diminished', 'minor', 'major', 'perfect', 'augmented'].map(value => ({ value }))}
                    name='intervalQuality'
                    value={state.intervalQuality}
                    onChange={intervalQuality => { onFormChange({ ...state, intervalQuality }) }}
                  />
                </Col>
                <Col>
                  <RadioButtons
                    values={Array.from(Array(8).keys()).map(it => it + 1).map(it => ({ value: `${it}` }))}
                    name='intervalSize'
                    value={state.intervalQuality}
                    onChange={intervalSize => { onFormChange({ ...state, intervalSize }) }}
                  />
                </Col>
              </Row>
            </Container>
          </Col>
          <Col>
            <KeySignatureSelector
              defaultValue={state.keySignature}
              onChange={keySignature => { onFormChange({ ...state, keySignature }) }}
            />
          </Col>
        </Row>
      </Form>
    </Container>
  )
}

export default IntervalSelector
