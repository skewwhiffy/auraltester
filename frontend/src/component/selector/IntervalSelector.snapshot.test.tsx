import IntervalSelector from './IntervalSelector'
import { Props as ClefSelectorProps } from '../../component/selector/ClefSelector'
import { Props as NoteSelectorProps } from '../../component/selector/NoteSelector'
import { Props as RadioButtonsProps } from '../../util/RadioButtons'
import renderer from 'react-test-renderer'

jest.mock('../../component/selector/ClefSelector', () => (props: ClefSelectorProps) => (
  <div>{JSON.stringify(props)}</div>
))

jest.mock('../../component/selector/NoteSelector', () => (props: NoteSelectorProps) => (
  <div>{JSON.stringify(props)}</div>
))

jest.mock('../../util/RadioButtons', () => (props: RadioButtonsProps) => (
  <div>{JSON.stringify(props)}</div>
))

describe('IntervalSelector snapshot', () => {
  it('matches snapshot', () => {
    const actual = renderer.create(
      <IntervalSelector onChange={() => { }} />
    )

    expect(actual).toMatchSnapshot()
  })
})