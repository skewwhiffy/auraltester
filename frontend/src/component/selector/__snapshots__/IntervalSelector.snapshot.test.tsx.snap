// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IntervalSelector snapshot matches snapshot 1`] = `
<div
  className="container"
>
  <form
    className=""
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <div>
          {}
        </div>
      </div>
      <div
        className="col"
      >
        <div>
          {"includeDoubleAccidentals":true}
        </div>
      </div>
      <div
        className="col"
      >
        <div
          className="container"
        >
          <div
            className="row"
          >
            <div
              className="col"
            >
              Interval
            </div>
          </div>
          <div
            className="row"
          >
            <div
              className="col"
            >
              <div>
                {"values":[{"value":"diminished"},{"value":"minor"},{"value":"major"},{"value":"perfect"},{"value":"augmented"}],"name":"intervalQuality"}
              </div>
            </div>
            <div
              className="col"
            >
              <div>
                {"values":[{"value":"1"},{"value":"2"},{"value":"3"},{"value":"4"},{"value":"5"},{"value":"6"},{"value":"7"},{"value":"8"}],"name":"intervalSize"}
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="col"
      >
        <div
          className="container"
        >
          <div
            className="row"
          >
            <div
              className="col"
            >
              Key signature
            </div>
          </div>
          <div
            className="row"
          >
            <div
              className="col"
            >
              <div>
                {"values":[{"value":"C#","label":"C# / A# minor"},{"value":"F#","label":"F# / D# minor"},{"value":"B","label":"B / G# minor"},{"value":"E","label":"E / C# minor"},{"value":"A","label":"A / F# minor"},{"value":"D","label":"D / B minor"},{"value":"G","label":"G / E minor"},{"value":"C","label":"None / C / A minor"},{"value":"F","label":"F / D minor"},{"value":"Bb","label":"Bb / G minor"},{"value":"Eb","label":"Eb / C minor"},{"value":"Ab","label":"Ab / F minor"},{"value":"Db","label":"Db / Bb minor"},{"value":"Gb","label":"Gb / Eb minor"},{"value":"Cb","label":"Cb / Ab minor"}],"name":"keySignature"}
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
